### PR TITLE
feat: check partition rule

### PR DESCRIPTION
**Pull request by waynexia**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
#3351

## What's changed and what's your intention?

Check the input partition rule set to make sure there is no cavity.

I'll file a follow-up PR to refactor the `partition` crate to remove the old implementation and install the new one directly (without a trait indirection).

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: src/partition/src/multi_dim.rs [hunk 1]
```diff
@@ -21,37 +21,46 @@ use serde::{Deserialize, Serialize};
 use snafu::ensure;
 use store_api::storage::RegionNumber;
 
-use crate::error::{self, Result};
+use crate::error::{self, ConjunctExprWithNonExprSnafu, Result, UnclosedValueSnafu};
 use crate::expr::{Operand, PartitionExpr, RestrictedOp};
 use crate::PartitionRule;
 
 #[derive(Debug, Serialize, Deserialize)]
```
Comment (killme2008) [inline src/partition/src/multi_dim.rs:28]: Please also add some docs for this struct.

File: src/partition/src/multi_dim.rs [hunk 2]
```diff
@@ -140,6 +149,132 @@ impl PartitionRule for MultiDimPartitionRule {
...
+                if let Operand::Expr(lhs) = expr.lhs.as_ref() {
+                    self.walk_expr(lhs)?
+                } else {
+                    ConjunctExprWithNonExprSnafu { expr: expr.clone() }.fail()?;
+                }
+                if let Operand::Expr(rhs) = expr.rhs.as_ref() {
+                    self.walk_expr(rhs)?
+                } else {
+                    ConjunctExprWithNonExprSnafu { expr: expr.clone() }.fail()?;
+                }
```
Comment (killme2008) [inline src/partition/src/multi_dim.rs:206]: Looks like this code can be refactored into:

```rust
match (expr.lhs.as_ref(), expr.rhs.as_ref()) {
    (Operand::Expr(lhs), Operand::Expr(rhs)) => {
        self.walk_expr(lhs)?;
        self.walk_expr(rhs)?
    },
    _ => ConjunctExprWithNonExprSnafu { expr: expr.clone() }.fail()?
}
```

File: src/partition/src/multi_dim.rs [hunk 3]
```diff
@@ -140,6 +149,132 @@ impl PartitionRule for MultiDimPartitionRule {
...
+            }
+            // Not conjunction
+            _ => {}
+        }
+
+        let (col, val) = match (expr.lhs.as_ref(), expr.rhs.as_ref()) {
+            (Operand::Expr(_), _)
+            | (_, Operand::Expr(_))
+            | (Operand::Column(_), Operand::Column(_))
+            | (Operand::Value(_), Operand::Value(_)) => todo!("invalid expr: {:?}", expr),
```
Comment (killme2008) [inline src/partition/src/multi_dim.rs:218]: It's better to return an error instead of panic.

File: src/partition/src/multi_dim.rs [hunk 4]
```diff
@@ -140,6 +149,132 @@ impl PartitionRule for MultiDimPartitionRule {
...
+                split_point.is_equal = true;
+            }
+            RestrictedOp::Gt => {
+                split_point.less_than_counter -= 1;
+            }
+            RestrictedOp::GtEq => {
+                split_point.less_than_counter -= 1;
+                split_point.is_equal = true;
+            }
+            RestrictedOp::And | RestrictedOp::Or => todo!("unreachable?"),
```
Comment (killme2008) [inline src/partition/src/multi_dim.rs:255]: What about adding a comment `already processed above`?

File: src/partition/src/multi_dim.rs [hunk 5]
```diff
@@ -198,4 +334,243 @@ mod tests {
...
+                PartitionExpr::new(
+                    Operand::Column("b".to_string()),
+                    RestrictedOp::GtEq,
+                    Operand::Value(datatypes::value::Value::String("s".into())),
+                ),
+            ],
+        );
+
+        // check rule
+        assert!(rule.is_err());
```
Comment (killme2008) [inline src/partition/src/multi_dim.rs:370]: Let's match the error, checking if it's expected.

File: src/partition/src/multi_dim.rs [hunk 6]
```diff
@@ -198,4 +334,243 @@ mod tests {
...
+                            RestrictedOp::GtEq,
+                            Operand::Value(datatypes::value::Value::Int64(20)),
+                        )),
+                    )),
+                ),
+            ],
+        );
+
+        // check rule
+        assert!(rule.is_err());
```
Comment (killme2008) [inline src/partition/src/multi_dim.rs:513]: The same suggestion as above. And for the following test assertions too.

File: src/partition/src/multi_dim.rs [hunk 7]
```diff
@@ -198,4 +346,297 @@ mod tests {
...
+
+        // check rule
+        assert_matches!(rule.unwrap_err(), Error::UnclosedValue { .. });
+    }
+
+    #[test]
+    fn duplicate_expr_case_1() {
+        // PARTITION ON COLUMNS (a) (
+        //     a <= 20,
+        //     b >= 10
```
Comment (fengjiachun) [inline src/partition/src/multi_dim.rs:586]: a or b?

**Review by fengjiachun — APPROVED** [review]
LGTM

**Review by killme2008 — APPROVED** [review]
LGTM

**Pull request merged**